clientapi/routing/register.go

```diff
@@ -225,17 +227,38 @@ func validateApplicationServiceUsername(username string) *util.JSONResponse {
 }
 
 // validatePassword returns an error response if the password is invalid
-func validatePassword(password string) *util.JSONResponse {
+func validatePassword(password string, cfg config.PasswordRequirements) *util.JSONResponse {
 	// https://github.com/matrix-org/synapse/blob/v0.20.0/synapse/rest/client/v2_alpha/register.py#L161
-	if len(password) > maxPasswordLength {
+	if len(password) > cfg.MaxPasswordLength {
 		return &util.JSONResponse{
 			Code: http.StatusBadRequest,
-			JSON: jsonerror.BadJSON(fmt.Sprintf("'password' >%d characters", maxPasswordLength)),
+			JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", cfg.MaxPasswordLength)),
 		}
-	} else if len(password) > 0 && len(password) < minPasswordLength {
+	} else if len(password) > 0 && len(password) < cfg.MinPasswordLength {
```

Comment:
This fails when `len(password) == 0` as it'll fallthrough.

Reply:
oh my, good catch. oops.

Reply:
hey, so i cant actually deny empty passwords here. the most recent failure for sytest on buildkite comes from the AS tests:

https://buildkite.com/matrix-dot-org/dendrite/builds/5406

It looks like that in synapse itself, we "allow" empty passwords here:

https://github.com/matrix-org/synapse/blob/642a42eddece60afbbd5e5a6659fa9b939238b4a/synapse/handlers/password_policy.py#L52-L60

From what I understand (and I'm still learning, so excuse any dumb assumptions :) ) is that Application Services actually create users (specifically around IRC bridging) that have empty passwords.

I _think_ we should be denying account registrations with empty passwords, but that seems like it might get a little weird with application services.

I could edit the sytest tests to send a password along, but I'm sort of afraid that by merging this in with checking with length 0 passwords, that it could break application services in the future that might do this. :\

I hope I'm understanding correctly and that I haven't word salad'd :laughing:

Reply:
We should allow empty passwords if it is an application service doing the registration, else we should reject.

Reply:
@kegsay i keep coming back to trying to get this one to pass, but my gorilla brain can't really figure out how to differentiate between registration attempts that are application services and ones that are not.

is there a header or something that is set?

i want to figure out a way how to do this without breaking backwards compatibility (saying "hey, you now need to set XYZ thing to say you are an application service") but im sort of barking up weird trees.

Reply:
[r.Type == authtypes.LoginTypeApplicationService](https://github.com/matrix-org/dendrite/blob/eadbfa7eea28beb5d706650ff9a158b68f37f521/clientapi/routing/register.go#L525) which could be passed to `validatePassword`. (Tested and now only `GET /register yields a set of flows` is still failing)

Reply:
Oh hell yeah! Okay, I'm coming off a busy weekend, but I'll push a new commit soon.

Thanks a ton for the tip! <3